### ISSUE TITLE
Some tests, e.g. test.html and test.jpg, still required access to the external https://download.moodle.org/unittest

### ISSUE THREAD

**Issue by scara** [moodlehq/moodle-docker#178]
Hello Everybody,
on updating my fork I see [an unrelated failure](https://github.com/scara/moodle-docker/runs/3234107458?check_suite_focus=true#step:5:41) on my local GHA run due to a temporary "external issue":
```
Debugging: cURL request for "https://download.moodle.org/unittest/test.html" failed, HTTP response code: HTTP/2 502 

* line 1682 of /lib/filelib.php: call to debugging()
* line 114 of /lib/tests/filelib_test.php: call to download_file_content()
* line 1526 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to core_filelib_testcase->test_download_file_content()
* line 1132 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to PHPUnit\Framework\TestCase->runTest()
* line 80 of /lib/phpunit/classes/advanced_testcase.php: call to PHPUnit\Framework\TestCase->runBare()
* line 722 of /vendor/phpunit/phpunit/src/Framework/TestResult.php: call to advanced_testcase->runBare()
* line 884 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to PHPUnit\Framework\TestResult->run()
* line 677 of /vendor/phpunit/phpunit/src/Framework/TestSuite.php: call to PHPUnit\Framework\TestCase->run()
* line 677 of /vendor/phpunit/phpunit/src/Framework/TestSuite.php: call to PHPUnit\Framework\TestSuite->run()
* line 677 of /vendor/phpunit/phpunit/src/Framework/TestSuite.php: call to PHPUnit\Framework\TestSuite->run()
* line 667 of /vendor/phpunit/phpunit/src/TextUI/TestRunner.php: call to PHPUnit\Framework\TestSuite->run()
* line 143 of /vendor/phpunit/phpunit/src/TextUI/Command.php: call to PHPUnit\TextUI\TestRunner->run()
* line 96 of /vendor/phpunit/phpunit/src/TextUI/Command.php: call to PHPUnit\TextUI\Command->run()
* line 61 of /vendor/phpunit/phpunit/phpunit: call to PHPUnit\TextUI\Command::main()
...
Time: 56:56.097, Memory: 1.03 GB

There was 1 failure:

1) core_filelib_testcase::test_download_file_content
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'47250a973d1b88d9445f94db4ef2c97a'
+'d41d8cd98f00b204e9800998ecf8427e'

/var/www/html/lib/tests/filelib_test.php:115
/var/www/html/lib/phpunit/classes/advanced_testcase.php:80
```
which is unexpected to me since I'd guess [`exttests`](https://github.com/moodlehq/moodle-exttests) should be targeted even here, in the local tests.

It appears that "a family of tests" in Moodle core code, based on `/lib/phpunit/classes/advanced_testcase.php::getExternalTestFileUrl()`, has been left out from the scope of `exttests`, probably due to the unavailability of a valid HTTPS certificate and the properly set of `TEST_EXTERNAL_FILES_HTTPS_URL`.

Should we, when running here the tests, define `TEST_EXTERNAL_FILES_HTTPS_URL` [as `false`](https://github.com/moodle/moodle/blob/dc437b51715eaee7e55998ff6eb52f751b19a306/lib/phpunit/classes/advanced_testcase.php#L628) to skip the above tests, avoiding to overload the Moodle Download site?

Note: I still miss why `getExternalTestFileUrl()` returns the HTTPS URL, since the local Moodle instance is exposed [via HTTP](https://github.com/moodlehq/moodle-docker/blob/f887c99383972136de09bac84603cc380d3c6a01/config.docker-template.php#L20) and the failed test [do not explicitly require HTTPS](https://github.com/moodle/moodle/blob/dc437b51715eaee7e55998ff6eb52f751b19a306/lib/tests/filelib_test.php#L87).

HTH,
Matteo

**Comment by stronk7** [comment]
Hi,

1. I think that the https ones can be put to http just specifying it also in config. For example, @ [moodle-ci-runner](https://github.com/moodlehq/moodle-ci-runner] we have both variables set to http ([link](https://github.com/moodlehq/moodle-ci-runner/blob/master/runner/master/config.template.php#L111-L114)). And it works perfectly. So I think that we can do pretty much the same here, in moodle-docker.

```
if (!empty(getenv('EXTTESTURL'))) {
    define('TEST_EXTERNAL_FILES_HTTP_URL', getenv('EXTTESTURL'));
    define('TEST_EXTERNAL_FILES_HTTPS_URL', getenv('EXTTESTURL'));
}
```

2. That said, yeah, we must ensure that there aren't any other calls to downloads (haven't looked) and fix all them.

3. Finally, curiously, this called my attention because, while I was playing to get multi-arch php moodlehq images (amd64 and arm64, yay, it seems to be [working great](https://hub.docker.com/repository/docker/stronk7/moodle-php-apache/tags?page=1&ordering=last_updated)), I've been performing a number or complete phpunit runs here (both architectures). And I've been getting a consistent failure with a test that downloads `user.jpg` (from memory), always with a CURL timeout error. But, the URL was the correct one (http://exttests/user.jpg), so I've to dig into that problem. It seems that other runs aren't affected, it only happens in my local docker instances (amd64 and arm64).

So, yeah, we should sort out 1 & 2. I'm going to start a run here without networking to see which tests are, effectively, trying to access outside (I imagine that there will be some H5P-related ones, but there shouldn't be others, agree).

**Comment by stronk7** [comment]
Here it's a complete run: https://pastebin.com/2fBj8czB

With only change performed being:

```
--- a/config.docker-template.php
+++ b/config.docker-template.php
@@ -47,6 +47,7 @@ $CFG->pathtophp = '/usr/local/bin/php';
 $CFG->phpunit_dataroot  = '/var/www/phpunitdata';
 $CFG->phpunit_prefix = 't_';
 define('TEST_EXTERNAL_FILES_HTTP_URL', 'http://exttests:9000');
+define('TEST_EXTERNAL_FILES_HTTPS_URL', 'http://exttests:9000');
 
 $CFG->behat_wwwroot   = 'http://webserver';
 $CFG->behat_dataroot  = '/var/www/behatdata';
```

And this is the list of problems (grouped by type) while it was run with networking disabled:

- core_iplookup_geoplugin_testcase::test_ipv4

- tool_mobile_external_testcase::test_get_public_config

- auth_oauth2_privacy_testcase::test_delete_data_for_users
- auth_oauth2_privacy_testcase::test_get_users_in_context
- auth_oauth2_privacy_testcase::test_delete_data_for_user
- auth_oauth2_privacy_testcase::test_delete_data_for_all_users_in_context
- auth_oauth2_privacy_testcase::test_export_user_data
- auth_oauth2_privacy_testcase::test_get_contexts_for_userid
- auth_oauth2_external_testcase::test_linked_logins
- auth_oauth2_external_testcase::test_clean_orphaned_linked_logins_with_issuer_id
- auth_oauth2_external_testcase::test_clean_orphaned_linked_logins
- core_oauth2_testcase::test_is_available_for_login
- core_oauth2_testcase::test_get_all_issuers
- core_oauth2_testcase::test_create_endpoints_for_standard_issuer with data set "IMS OBv2.1 without slash in baseurl should work too" ('imsobv2p1', '.well-known/badgeconnect.json', false, 'https://dc.imsglobal.org')
- core_oauth2_testcase::test_create_endpoints_for_standard_issuer with data set "IMS OBv2.1" ('imsobv2p1', '.well-known/badgeconnect.json', false, 'https://dc.imsglobal.org/')
- core_oauth2_testcase::test_create_endpoints_for_standard_issuer with data set "Google will work too with a valid baseurl parameter" ('google', '.well-known/openid-configuration', true, 'https://accounts.google.com/')
- core_oauth2_testcase::test_create_endpoints_for_standard_issuer with data set "Google" ('google', '.well-known/openid-configuration')
- core_oauth2_testcase::test_create_and_delete_standard_issuers

- core_googlelib_testcase::test_invalid_google_api_key

- core_antivirus_testcase::test_manager_none_quarantine_data_virus
- core_antivirus_testcase::test_manager_quarantine_data_virus
- core_antivirus_testcase::test_manager_send_message_to_admin_email_scan_data_virus
- core_antivirus_testcase::test_manager_send_message_to_user_email_scan_data_virus
- core_antivirus_testcase::test_manager_scan_data_virus
- core_antivirus_testcase::test_manager_scan_data_error
- core_antivirus_testcase::test_manager_none_quarantine_file_virus
- core_antivirus_testcase::test_manager_quarantine_file_virus
- core_antivirus_testcase::test_manager_send_message_to_admin_email_scan_file_virus
- core_antivirus_testcase::test_manager_send_message_to_user_email_scan_file_virus
- core_antivirus_testcase::test_manager_scan_file_virus
- core_antivirus_testcase::test_manager_scan_file_error

Basically all the oauth ones are about resolving some host, and the antivirus ones about (surprisingly) performing ip-lookups. Nothing related with exttests, so far.

There is also a failure that I'm getting locally all the time since some days ago and cannot reproduce @ CIs, but that's another story, heh.

So, maybe, all we need is to add `TEST_EXTERNAL_FILES_HTTPS_URL` and done.

**Comment by scara** [comment]
Hi @stronk7,
thanks a ton for your deep diving! 👍 

My HTTPS failure was related to https://github.com/moodle/moodle/blob/0d0e66d37c6271657d76f948db9eeb10c139e7ae/lib/tests/filelib_test.php#L111-L115 so it doesn't hurt even there.

+1 to your proposal to mimic what already in place at https://github.com/moodlehq/moodle-ci-runner.

Unfortunately, we can define and distribute the `$CFG->geoip2file` file to avoid calling `http://www.geoplugin.net/json.gp?ip=<IP Address>`.

TIA,
Matteo
